lib/hal/portable/st/aws_hal_perf_cntr.c

```diff
@@ -0,0 +1,233 @@
+/**
+ * @brief Implement overflow ISR for TIM5
+ */
+void HAL_TIM5_PeriodElapsedCallback(TIM_HandleTypeDef *htim)
+{
+    ullTimerOverflow++;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief TIM5 ISR
+ *
+ * @note a cleaner way to handle timer IRQ for ST is to implement a branch for
+ * .Instance == TIM5 in HAL_TIM_PeriodElapsedCallback(). Yet, to not expose static
+ * variable ullTimerOverflow, enabling USE_HAL_TIM_REGISTER_CALLBACKS and use
+ * .PeriodElapsedCallback for TIM5.
+ */
+
+void TIM5_IRQHandler(void)
+{
+    HAL_TIM_IRQHandler( &xTimerHandle );
+}
+/*-----------------------------------------------------------*/
+
+void aws_hal_perfcounter_open(void)
+{
+    uint32_t ulTimClock = 0;
+
+    /* Clear static variable which tracks total number of timer overflow. */
+    ullTimerOverflow = 0;
+
+    /* Get timer clock frequency. */
+    ulTimClock = HAL_RCC_GetPCLK1Freq();
+
+    /* Check user input frequency is valid. */
+    configASSERT( HAL_PERF_COUNTER_FREQ <= ulTimClock );
+
+    /* From STM32L475xx chip datasheet:
+     * - TIM2 and TIM5 are 32-bit general purpose timers, while others are all 16-bit.
+     * - Peripheral current consumption (Range 1/Range 2/Low power and sleep):
+     *   TIM2: 6.8/6.5/6.3 (uA/MHZ)
+     *   TIM5: 6.5/5.5/6.1 (uA/MHZ)
+     * To minimize the need for handling timer overflow interrupt with relative low cost,
+     * prefer to use TIM5 as the hardware timer for performance counter.
+     */
+    xTimerHandle.Instance = TIM5;
+
+    /* To minimize the need for serving overflow IRQ,
+     * .Period is always set to the max value the counter allows.
+     * Performance counter resolution is adjusted with .Prescaler.
+     */
+    xTimerHandle.Init.Prescaler = ulTimClock / HAL_PERF_COUNTER_FREQ;
+    xTimerHandle.Init.CounterMode = TIM_COUNTERMODE_UP;
+    xTimerHandle.Init.Period = HW_TIMER_32_CONST_PERIOD;
+    xTimerHandle.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
+    xTimerHandle.Init.RepetitionCounter = 0x00;
+    xTimerHandle.Init.AutoReloadPreload = TIM_AUTORELOAD_PRELOAD_ENABLE;
+
+    xTimerHandle.Channel = HAL_TIM_ACTIVE_CHANNEL_CLEARED;
+
+    xTimerHandle.Lock = HAL_UNLOCKED;
+
+    xTimerHandle.State = HAL_TIM_STATE_RESET;
+
+    HAL_TIM_Base_Init( &xTimerHandle );
+
+    /* Only generate interrupt when overflow. */
+    __HAL_TIM_URS_ENABLE( &xTimerHandle );
+
+    /* Clear the initial SR.UIF bit before timer starts. */
+    __HAL_TIM_CLEAR_FLAG( &xTimerHandle, TIM_FLAG_UPDATE);
+
+    /* Register overflow IRQ callback. */
+    HAL_TIM_RegisterCallback( &xTimerHandle, HAL_TIM_PERIOD_ELAPSED_CB_ID, HAL_TIM5_PeriodElapsedCallback );
+
+    /* Timer and interrupt are configured. Now start timer. */
+    HAL_TIM_Base_Start_IT( &xTimerHandle );
+}
+
+/*-----------------------------------------------------------*/
+
+void aws_hal_perfcounter_close( void)
+{
+    HAL_TIM_Base_Stop_IT( &xTimerHandle );
+
+    HAL_TIM_Base_DeInit( &xTimerHandle );
```

Comment:
Should this be HAL_TIM_Base_MspDeInit?

Reply:
HAL_TIM_Base_{Init, DeInit} call HAL_TIM_Base_Msp{Init, DeInit} internally. From ST HAL programming manual, *_Msp*() is pretty much an extension point where user can define additional routine for initialization and deinitialization. So, in our case, besides regular timer initialization routine, we are turning on interrupt for TIM5.